credentials/tls/certprovider/store.go

```diff
@@ -80,9 +80,19 @@ func GetProvider(name string, config interface{}, opts Options) (Provider, error
 	if builder == nil {
 		return nil, fmt.Errorf("no registered builder for provider name: %s", name)
 	}
-	stableConfig, err := builder.ParseConfig(config)
-	if err != nil {
-		return nil, err
+
+	var (
+		stableConfig StableConfig
+		err          error
+	)
+	if c, ok := config.(StableConfig); ok {
+		// The config passed to the store has already been parsed.
```

Comment:
> The other option would be to modify the store to accept config of type certprovider.StableConfig and not interface{}.

Argh.  Either way is a bit awkward:

1. Validate the config (by parsing it), then create the provider (parsing the config again with redundant yet necessary error handling).

2. Parse the config (using a Builder found by looking up the name), then use the StableConfig to create the provider (using a Builder found again by looking up the name, later, with a chance to mismatch the StableConfig).

Also, both of these choices leaves me wondering where "Options" should go -- if it's options for the provider, shouldn't it be passed to `ParseConfig` so it can fail with a relevant error if it doesn't support the options given?

What if the StableConfig contained the provider name so there wouldn't be a chance for a mismatch?  Or, instead of even having a StableConfig*, make the parser produce a Builder?

\* - We'd still need a StableConfig for the Store, but not for users.

Reply:
I like the option of having `StableConfig` contain the provider name. This seems to be a safe thing to do along with changing the `store` to accept `StableConfig` instead of `interface{}`.

The problem with having the `Parser` produce a `Builder` is that we already want the latter to perform some amount of caching/sharing of common resources used by instances of the same plugin. For example, the MeshCA `Builder` is the one which owns the `ClientConn` to the MeshCA server, and this is shared between multiple instances of the plugin which all talk to the same server.

And wrt the `Options`, I like to think of them like the options passed to a balancer (which is done at `Build` time). And given that the `Options` struct is common to all plugins, we probably wouldn't be adding much to it and it simply contains the certName and whether the plugin should fetch root or identity certs. There could possibly be plugins for whom a given config validity also depends on the given options, but that seems to be rare and we could possibly change the `Build` call to return an `error` if we want to support these scenarios.

Reply:
> This sharing could be moved to the parser instead. I don't think this would be hard.

Sure. But logically doesn't seem right for the `Parser` to be burdened with that functionality.

> Do all plugins need to support both root & identity certs?

All the plugins that we've seen so far do support getting both root & identity certs. In some cases, the caller is not interested in identity certs (TLS vs mTLS). It would be a little strange for a plugin to support only one type of certificate. But there is nothing that prevents one from writing a plugin which supports only one type of cert.

> If no, what do we want to happen if the wrong type is requested from a provider which only supports one?

In this case, I would expect the call to `Build()` to fail. But that call doesn't return an error, it can only return a `nil` Provider. When we discussed whether or not to return an error from `Build`, we probably did not discuss this case. We only discussed the case of the config being wrong, and we decided that since we have a separate `Parse` method, and since `Build` is taking the result of `Parse`, config parsing should not fail at build time.

> Do we not know at config validation time which will be needed?

With the current design where config parsing is done at bootstrap time, we do not know.

Reply:
> With the current design where config parsing is done at bootstrap time, we do not know.

In this case we need to pass `Options` to `Build`.  `Build` should really return an error instead of just a `nil` when it's unhappy.

> But logically doesn't seem right for the Parser to be burdened with that functionality.

The alternative flow is a bit weird:

1. Pass config to ParseConfig; get back StableConfig containing plug-in name.
2. Use StableConfig plug-in name to find the Builder again; pass this same StableConfig to Builder.Build.

It would be more streamlined if the StableConfig could just produce the provider itself.

Reply:
I will make the following changes in a follow-up PR:
1. Change `Build()` to return `(Provider, error)`
2. Change the `StableConfig` interface to also have a `Name()` method
3. Change `certprovider.GetProvider()` to accept config of type `StableConfig` instead of `interface{}`. 

The flow for the xDS (with store) case will be as follows:
1. During bootstrap, we read JSON configs and call `Parse()` and store the returned `StableConfig`
2. At runtime, when CDS balancer receives security config, we lookup the appropriate `StableConfig` and call `GetProvider()`.
This doesn't look too weird. And given that we will have `Options` only at runtime, this seems more appropriate.

For the non-store case, the flow will be as follows (if the plugin doesn't provide a ctor):
1. Call `Parse()` and get `StableConfig`.
2. Pass the returned `StableConfig` back to `Build()` and get a `Provider`.
This is not super ideal, but we will not really have use-cases like this because plugins who expect to be used outside of the store should ideally provide a direct constructor.

Wdyt?

Reply:
> This doesn't look too weird
> This is not super ideal

The API is still experimental and we should design it to look how we want given the way it will be used.  Your statements aren't really ringing endorsements.

Wouldn't it make more sense, flow-wise, to add a method to `StableConfig` to create the `Provider`, moving it there from the `Builder`?  That would be more linear: `Builder->StableConfig->Provider` instead of `Builder->StableConfig;Builder+StableConfig->Provider`.  Ignoring the names for now, this is a leaner API structure -- fewer methods to implement, fewer calls to get a provider, less redundancy (and chance for mistakes), same functionality.  We would want to change the names, though.

Agree that any changes here can be done in a follow-up PR.

Reply:
I think I see your point now. I can imagine an API which would be along the following lines:

```go
package certprovider

func Register(p Parser) {}

type Parser interface {
  Parse(interface{}) (Config, error)
  Name() string
}

type Config interface {
  Name() string
  Canonical() []byte
  Build(Options) (Provider, error) // Still needs to return error since options might be incompatible.
}

func GetProvider(cfg Config, o Options) (Provider, error) {
  // storeKey: {cfg.Name(), cfg.Canonical(), o}
  // Check for existing provider with above key. If found, incr refCnt and return
  // Else, call cfg.Build(o)
  // Add newly returned provider to store and return it
}
```

The flow here with the store will be as follows:
- At bootstrap time, call `Parse()` and pass it the JSON configs. Cache the returned `Config` type.
- At CDS LB, call `certprovider.GetProvider(cfg, options)`

The flow without a store would be:
- Call `Parse()` and pass it the config of whatever type
- Call `cfg.Build()` and get a `Provider`.

Definitely fewer calls, fewer methods to implement.

Reply:
The goal is to minimize the things the user can do wrong with the API, with the top priority being consumers and the second priority being implementers.

- Needing `Name` and `Canonical` in `Config` is not ideal, as they are visible to consumers but not needed except by the store.  `Name` is especially confusing, since it's redundant with `Parser.Name()`.  This is one reason why I suggested that type to be a struct earlier:

```go
type Config struct {}
func NewConfig(name string, canonical []byte, builder func(Options) Provider) *Config
func (*Config) Build(Options) (Provider, error) // does store functionality
```

Here `NewConfig` would capture the name and canonical bytes, and the user would never see them.

- `GetProvider` shouldn't be necessary no matter which way we go.  We can wrap the `Config` with a `storeConfig` (or whatever) that does the map lookup at `Build` time.  Otherwise, the user is getting a `Build` function in `Config`, yet they're not supposed to call it.  Another chance to make a mistake.  But we do need a `ParseConfig(name string, cfg interface{}) (Config, error)` function.

- Do we need a `ParseOptions` (renaming `Options` to `BuildOptions`)?

Reply:
Does this look closer to what you have in mind?

```go
package certprovider

func Register(p Parser) {}

type Parser interface {
  Parse(interface{}) (*Config, error)
  Name() string
}

type Config struct {
  name string
  canonicalCfg []byte
  builder func(BuildOptions) Provider

  store *store // Reference to the package global store
}

// storeKey, wrappedProvider, store, provStore remain unchanged.

func NewConfig(name string, canonical []byte, builder func(Options) Provider) *Config {}

func (c *Config) Build(BuildOptions) (Provider, error) {
  // make a storeKey with {name, canonical, opts}
  // if key found in store, increment refCnt and return
  // else, use `builder` function to create a new provider, add it to the store and return it.
}
```

```go

package fooPlugin

func init() {
  certprovider.Register(fooBuilder)
}

type fooBuilder struct {
  // Shared stuff between providers.
}

func (fooBuilder) Name() {}

func (fooBuilder) Parse(cfg interface{}) (*Config, error) {
  // Parse the given config, and error out if required
  // Create a fooProvider, initialize it with common stuff, and the passed in config
  // return certprovider.NewConfig(fooName, parsed_canonical_config, fp.buildFoo())
}

type fooProvider struct {
  // common stuff
  // config

  // implements Provider interface
}

func (fp *fooProvider) buildFoo(o certprovider.BuildOptions) (Provider, error) {
  // use common stuff and config to initialize self
  // return self
}

// Or I guess, if the `buildFoo` is implemented as a closure inside `Parse`, things will be cleaner and simpler.

Reply:
Yes, I think that's what I proposed above, but with different names (I used `Builder` instead of `Config`).  :)

Regarding names...  This might be a radical change, but what about: `Provider.ParseConfig()` -> `KeyReaderBuilder.Build()` -> `KeyReader.Read()`.  Is this more in line with C/Java terminology?

`buildFoo` can either be a closure or a method; whatever is simpler for the plugin.

Do we need a `ParseOptions` for future-proofing?  (Did we already decide we didn't?)

All this stuff can wait.  I'll review the rest of the PR now.

Reply:
> Yes, I think that's what I proposed above, but with different names (I used `Builder` instead of `Config`). :)

I guessed that would be coming :)

> Do we need a ParseOptions for future-proofing? (Did we already decide we didn't?)

I don't see a need for it at this point, but we could add an empty struct, and decide what to do with it when we remove the experimental tag on the package.

> All this stuff can wait. I'll review the rest of the PR now.

👍

Reply:
> Regarding names... This might be a radical change, but what about: Provider.ParseConfig() -> KeyReaderBuilder.Build() -> KeyReader.Read(). Is this more in line with C/Java terminology?

I don't remember their API now. Will check.